### PR TITLE
Enable windowHistorySupport by default

### PR DESCRIPTION
**Pull request by timneutkens**
## What?

Enables `experimental.windowHistorySupport` by default. It has been in experimental for quite a while now and has been successfully dogfooded in vercel.com as well.

Closes NEXT-2076
Closes NEXT-1773

### CODE REVIEW

**Comment by timneutkens** [comment]
<b>In this stack:</b> <code>🔧 Enable windowHistorySupport by default, improve user experience</code><a href="https://withgraphite.notion.site/Graphite-beta-features-EXTERNAL-496843566b204c7a97a2a21f8dea5434">(?)</a>

* `canary`
  * **#60557** <a href="https://app.graphite.dev/github/pr/vercel/next.js/60557?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @timneutkens and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 01-12-Enable_windowHistorySupport_by_default | Change |
| - | - | - | - |
| buildDuration | 12.9s | 12.8s | N/A |
| buildDurationCached | 7s | 6s | N/A |
| nodeModulesSize | 200 MB | 200 MB | N/A |
| nextStartRea..uration (ms) | 424ms | 433ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js 01-12-Enable_windowHistorySupport_by_default | Change |
| - | - | - | - |
| 193.HASH.js gzip | 181 B | 182 B | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | N/A |
| 433-HASH.js gzip | 28.6 kB | 28.7 kB | ⚠️ +186 B |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 239 B | 242 B | N/A |
| main-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 73.8 kB | 74 kB | ⚠️ +186 B |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 01-12-Enable_windowHistorySupport_by_default | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 01-12-Enable_windowHistorySupport_by_default | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 183 B | 181 B | N/A |
| amp-HASH.js gzip | 504 B | 502 B | N/A |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 253 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.28 kB | 4.28 kB | N/A |
| index-HASH.js gzip | 255 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | ✓ |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.4 kB | 3.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 01-12-Enable_windowHistorySupport_by_default | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 01-12-Enable_windowHistorySupport_by_default | Change |
| - | - | - | - |
| index.html gzip | 529 B | 527 B | N/A |
| link.html gzip | 541 B | 540 B | N/A |
| withRouter.html gzip | 524 B | 522 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js 01-12-Enable_windowHistorySupport_by_default | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.8 kB | 93.8 kB | N/A |
| page.js gzip | 148 kB | 148 kB | ⚠️ +159 B |
| Overall change | 148 kB | 148 kB | ⚠️ +159 B |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 01-12-Enable_windowHistorySupport_by_default | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 623 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.5 kB | 37.5 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.07 kB | 2.07 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 01-12-Enable_windowHistorySupport_by_default | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 169 kB | 169 kB | ✓ |
| app-page-exp..prod.js gzip | 95.2 kB | 95.2 kB | ✓ |
| app-page-tur..prod.js gzip | 95.9 kB | 95.9 kB | ✓ |
| app-page-tur..prod.js gzip | 90.5 kB | 90.5 kB | ✓ |
| app-page.run...dev.js gzip | 142 kB | 142 kB | ✓ |
| app-page.run..prod.js gzip | 89.8 kB | 89.8 kB | ✓ |
| app-route-ex...dev.js gzip | 24.1 kB | 24.1 kB | ✓ |
| app-route-ex..prod.js gzip | 16.7 kB | 16.7 kB | ✓ |
| app-route-tu..prod.js gzip | 16.7 kB | 16.7 kB | ✓ |
| app-route-tu..prod.js gzip | 16.3 kB | 16.3 kB | ✓ |
| app-route.ru...dev.js gzip | 23.5 kB | 23.5 kB | ✓ |
| app-route.ru..prod.js gzip | 16.3 kB | 16.3 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.38 kB | 9.38 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.65 kB | 9.65 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| pages.runtim...dev.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| server.runti..prod.js gzip | 49.6 kB | 49.6 kB | ✓ |
| Overall change | 940 kB | 940 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>433-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: ae659e10c34ea0a8ee1776c9864e6faa93469e73

**Comment by ijjk** [comment]
## Failing test suites 
Commit: ae659e10c34ea0a8ee1776c9864e6faa93469e73

`pnpm test test/integration/edge-runtime-module-errors/test/module-imports.test.js` 
- Edge runtime code with imports > Edge API statically importing node.js module > throws unsupported module error in dev at runtime and highlights the faulty line

<details>
<summary>Expand output</summary>

  ● Edge runtime code with imports › Edge API statically importing node.js module › throws unsupported module error in dev at runtime and highlights the faulty line

    expect(received).toBe(expected) // Object.is equality

    Expected: 500
    Received: 404

      86 |       context.app = await launchApp(context.appDir, context.appPort, appOption)
      87 |       const res = await fetchViaHTTP(context.appPort, url)
    > 88 |       expect(res.status).toBe(500)
         |                          ^
      89 |       await check(async () => {
      90 |         expectUnsupportedModuleDevError(
      91 |           moduleName,

      at Object.toBe (integration/edge-runtime-module-errors/test/module-imports.test.js:88:26)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by huozhi — APPROVED** [review]

File: packages/next/src/client/components/app-router.tsx [hunk 1]
```diff
@@ -467,65 +464,64 @@ function Router({
...
+    // Ensure the canonical URL in the Next.js Router is updated when the URL is changed so that `usePathname` and `useSearchParams` hold the pushed values.
+    const applyUrlFromHistoryPushReplace = (
+      url: string | URL | null | undefined
+    ) => {
+      const href = window.location.href
+      startTransition(() => {
+        dispatch({
+          type: ACTION_RESTORE,
+          url: new URL(url ?? href, href),
+          tree: window.history.state.__PRIVATE_NEXTJS_INTERNALS_TREE,
```
Comment (greysteil) [inline packages/next/src/client/components/app-router.tsx:477]: @timneutkens after upgrading from `14.0.4` to `14.1.0` I'm seeing errors from this line when using Microsoft's `@azure/msal-browser` with NextJS. Specifically the error is:

`null is not an object (evaluating 'window.history.state.__PRIVATE_NEXTJS_INTERNALS_TREE')`

The flow that causes the error is that Microsoft has just redirected a user to my site with a hash in the URL, and the client code on my site (via the `@azure/msal-browser` package) is trying to remove the hash from the URL. Specifically, this line is being called:

https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/2d4d0b2cde8e072ba6c64aea9b251984b2dd4021/lib/msal-browser/src/utils/BrowserUtils.ts#L21-L25

I have a patch locally that "fixes" this by adding `if (window.history.state == null) return;` directly above `startTransition(() => { ...` a couple of lines above here, but I don't understand Next.js's internals well enough to know if that's the right fix. I can submit a PR with it if you think it is.

❤️ for all your work on Next.js.

Comment (greysteil) [reply packages/next/src/client/components/app-router.tsx:477]: @timneutkens gentle nudge on ☝️

**Pull request merged**